cpp/src/arrow/array/diff.cc

```diff
@@ -0,0 +1,737 @@
+    insert_.resize(end(edit_count_ + 1), false);
+
+    // try deleting from base first
+    for (uint64_t i = begin(edit_count_), i_out = end(edit_count_); i != end(edit_count_);
+         ++i, ++i_out) {
+      endpoint_base_[i_out] = ExtendFrom(DeleteOne(GetEditPoint(edit_count_, i))).base;
+    }
+
+    // check if inserting from target could do better
+    for (uint64_t i = begin(edit_count_), i_out = end(edit_count_) + 1;
+         i != end(edit_count_); ++i, ++i_out) {
+      auto x_endpoint = GetEditPoint(edit_count_ + 1, i_out);
+
+      endpoint_base_[i_out] = ExtendFrom(InsertOne(GetEditPoint(edit_count_, i))).base;
+      auto y_endpoint = GetEditPoint(edit_count_ + 1, i_out);
+
+      if (y_endpoint.base - x_endpoint.base >= 0) {
+        insert_[i_out] = true;
+      } else {
+        endpoint_base_[i_out] = x_endpoint.base;
+      }
+    }
+
+    ++edit_count_;
+
+    // check for completion
+    EditPoint finish = {base_end_, target_end_};
+    for (uint64_t i = begin(edit_count_); i != end(edit_count_); ++i) {
+      if (GetEditPoint(edit_count_, i) == finish) {
+        finish_index_ = i;
+        return;
+      }
+    }
+  }
+
+  bool Done() { return finish_index_ != static_cast<uint64_t>(-1); }
+
+  Status GetEdits(MemoryPool* pool, std::shared_ptr<Array>* out) {
+    DCHECK(Done());
+
+    int64_t length = edit_count_ + 1;
+    std::shared_ptr<Buffer> insert_buf, run_length_buf;
+    RETURN_NOT_OK(AllocateBitmap(pool, length, &insert_buf));
+    RETURN_NOT_OK(AllocateBuffer(pool, length * sizeof(uint64_t), &run_length_buf));
+    auto run_length = reinterpret_cast<uint64_t*>(run_length_buf->mutable_data());
+
+    auto index = finish_index_;
+    auto endpoint = GetEditPoint(edit_count_, finish_index_);
+
+    for (int64_t i = edit_count_; i > 0; --i) {
+      bool insert = insert_[index];
+      BitUtil::SetBitTo(insert_buf->mutable_data(), i, insert);
+
+      auto x_minus_y = (endpoint.base - base_begin_) - (endpoint.target - target_begin_);
+      if (insert) {
+        ++x_minus_y;
+      } else {
+        --x_minus_y;
+      }
+      index = (i - 1 - x_minus_y) / 2 + begin(i - 1);
+
+      // endpoint of previous edit
+      auto previous = GetEditPoint(i - 1, index);
+      run_length[i] = endpoint.base - previous.base - !insert;
+
+      endpoint = previous;
+    }
+    BitUtil::SetBitTo(insert_buf->mutable_data(), 0, false);
+    run_length[0] = endpoint.base - base_begin_;
+
+    ARROW_ASSIGN_OR_RAISE(
+        *out,
+        StructArray::Make({std::make_shared<BooleanArray>(length, insert_buf),
+                           std::make_shared<UInt64Array>(length, run_length_buf)},
+                          {field("insert", boolean()), field("run_length", uint64())}));
+    return Status::OK();
+  }
+
+ private:
+  uint64_t finish_index_ = -1;
+  uint64_t edit_count_ = 0;
+  Iterator base_begin_, base_end_;
+  Iterator target_begin_, target_end_;
+  std::vector<Iterator> endpoint_base_;
```

Comment:
Can you comment on what this is? Is "base" related to the "base" array? Otherwise perhaps rename...